File: app/src/main/java/com/chebuso/chargetimer/Factory.java
package com.chebuso.chargetimer;

import android.app.Activity;
import android.content.Context;

import com.chebuso.chargetimer.notifications.IResourceProvider;
import com.chebuso.chargetimer.notifications.NotificationScheduler;
import com.chebuso.chargetimer.notifications.ResourceProvider;
import com.chebuso.chargetimer.settings.ISettingsReader;
import com.chebuso.chargetimer.settings.ISettingsWriter;
import com.chebuso.chargetimer.settings.SharedPreferenceSettingsWriter;
import com.chebuso.chargetimer.settings.SharedPreferenceSettingsReader;

class Factory {

    static ISettingsReader createSettingsReader(Context context){
        return new SharedPreferenceSettingsReader(context);
    }

    static ISettingsWriter createSettingsWriter(Context context) {
        return new SharedPreferenceSettingsWriter(context);
    }

    static NotificationScheduler createScheduler(Activity activity){
        IResourceProvider resourceProvider = new ResourceProvider(activity);
        ISettingsWriter settingsWriter = new SharedPreferenceSettingsWriter(activity);
        ISettingsReader settingsProvider = createSettingsReader(activity);

        return new NotificationScheduler(activity, settingsProvider, resourceProvider, settingsWriter);
    }
}

File: EvChargeTimer/app/src/main/java/com/chebuso/chargetimer/settings/Factory.kt
package com.chebuso.chargetimer.settings

import android.app.Activity
import android.content.Context
import com.chebuso.chargetimer.notifications.NotificationScheduler
import com.chebuso.chargetimer.notifications.ResourceProvider


object Factory {
    fun createSettingsReader(context: Context): ISettingsReader =
        SharedPreferenceSettingsReader(context)

    fun createSettingsWriter(context: Context): ISettingsWriter =
        SharedPreferenceSettingsWriter(context)

    fun createScheduler(activity: Activity): NotificationScheduler {
        val settingsProvider = createSettingsReader(activity)
        val settingsWriter = createSettingsWriter(activity)
        val resourceProvider = ResourceProvider(activity)
        return NotificationScheduler(activity, settingsProvider, settingsWriter, resourceProvider)
    }
}